lib/unison-sqlite/src/Unison/Sqlite/Connection.hs

```diff
@@ -121,51 +131,96 @@ closeConnection (Connection _ _ conn) =
 
 execute :: Sqlite.ToRow a => Connection -> Sql -> a -> IO ()
 execute conn@(Connection _ _ conn0) s params =
-  Sqlite.execute conn0 (coerce s) params `catch` \(exception :: Sqlite.SQLError) ->
+  doExecute `catch` \(exception :: Sqlite.SQLError) ->
     throwSqliteQueryException
       SqliteQueryExceptionInfo
         { connection = conn,
           exception = SomeSqliteExceptionReason exception,
           params = Just params,
           sql = s
         }
+  where
+    doExecute =
+      if debugTraceQueries
+        then do
+          Text.hPutStrLn stderr ("query:  " <> coerce s)
+          Text.hPutStrLn stderr ("params: " <> Text.pack (anythingToString params))
+          Text.hPutStrLn stderr "----------"
+          run
+        else run
+      where
+        run = Sqlite.execute conn0 (coerce s) params
 
 executeMany :: Sqlite.ToRow a => Connection -> Sql -> [a] -> IO ()
 executeMany conn@(Connection _ _ conn0) s params =
-  Sqlite.executeMany conn0 (coerce s) params `catch` \(exception :: Sqlite.SQLError) ->
+  doExecuteMany `catch` \(exception :: Sqlite.SQLError) ->
     throwSqliteQueryException
       SqliteQueryExceptionInfo
         { connection = conn,
           exception = SomeSqliteExceptionReason exception,
           params = Just params,
           sql = s
         }
+  where
+    doExecuteMany =
+      if debugTraceQueries
+        then do
+          Text.hPutStrLn stderr ("query:  " <> coerce s)
```

Comment:
Rather than duplicating this block in every execute function couldn't we pull it out into:

```
debugQuery :: Sql -> [a] -> IO ()
debugQuery = do
  when debugTraceQueries $ do
    ...
```

Then do:

```haskell
executeMany conn@(Connection _ _ conn0) s params = do
  debugQuery s params
  doExecuteMany `catch` \(exception :: Sqlite.SQLError) ->
    ...
```

?

Seems like less duplication, and I think it'd slightly simply the control flow too.

Reply:
I don't really understand the suggestion, sorry

Reply:
The idea is to pull out a top-level `debugQuery` helper rather than define it 3x in where clauses.
Just seems like we're duplicating our debug logic in every function when we could share it; it's nice to be able to change it in one place and have it affect every usage site at once. Sure, there's the location without any params, but passing `[]` is fine there.

Not a huge deal either way, but seems both simpler and DRY'er to me :shrug:

Reply:
Oh, I see. Well, `execute` and `executeMany` also don't return results, so there's more than just the params difference. I can move some code around to use `when`, though, that seems like an improvement.